### PR TITLE
[1.1.x] Fix G28 Bug #8392

### PR DESCRIPTION
**Pull request by RowanMeara**
When homing to Z_MIN, movements can occur in the +Z direction that are much larger than Z_HOMING_HEIGHT under the circumstances described in #8392  and below.

Currently if a user turns their printer on, moves to a Z position less than 0 (eg Z=-20), and homes their z-axis, the z-axis will move to Z=Z_HOMING_HEIGHT first.  For instance, if Z=-20 and Z_HOMING_HEIGHT=5, a 25mm move will be performed to Z=5 at the start of G28.  This is an issue if the z-axis is at its maximum position when the printer is turned on because it will crash. #8392 

This pull request changes the G28 behavior so that the Z axis will never be raised by more than Z_HOMING_HEIGHT distance before beginning the homing move.

Edit: There seems to be some confusion on what this fixes so I'll try and give an example with a 200x200x200 printer and a Z_HOMING_HEIGHT of 5.

Here is what happens currently:

| Action | Physical Position | Where the Printer Thinks it is |
| :-------|:-------------------:|--------------------------------:|
| Turn the Printer On | Z=200 | Z=0 | 
| G1 Z-50 | Z=150 | Z=-50 |
| G28 (raise Z to safehome) | Z = 200 + Z_HOMING_HEIGHT |  Z = 0 + HOMING_HEIGHT |
| The printer has crashed 

Here is what I think should happen:

| Action | Physical Position | Where the Printer Thinks it is |
| :-------|:-------------------:|--------------------------------:|
| Turn the Printer On | Z=200 | Z=0 | 
| G1 Z-50 | Z=150 | Z=-50 |
| G28 (raise Z to safehome) | Z = 150 + Z_HOMING_HEIGHT |  Z = -50 + HOMING_HEIGHT |
| G28 (home x and y) | Z = 155 | Z = -45 | 
| G28 (home z) | Z = 0 | Z = 0 |

### CODE REVIEW

File: Marlin/Marlin_main.cpp [hunk 1]
```diff
@@ -4087,14 +4087,21 @@ inline void gcode_G28(const bool always_home_all) {
 
       if (home_all || homeX || homeY) {
         // Raise Z before homing any other axes and z is not already high enough (never lower z)
-        destination[Z_AXIS] = Z_HOMING_HEIGHT;
-        if (destination[Z_AXIS] > current_position[Z_AXIS]) {
+        // We should never move more than Z_HOMING_HEIGHT distance.
+        #if DISABLED(MIN_SOFTWARE_ENDSTOP_Z)
+          if (current_position[Z_AXIS] < 0.0f)
```
Comment (AnHardt) [inline Marlin/Marlin_main.cpp:4092]: What is the relevance of SOFTWARE_ENDSTOP? Before homing they can't work and should not be active for that. And after once homed the problem is gone.

Comment (RowanMeara) [reply Marlin/Marlin_main.cpp:4092]: Software endstops work before homing.  

 When the printer is turned on, the zero position is set to the current position of the printer.  Software endstops behave accordingly.  You can test this by turning on your printer and trying a G1 X-10.  Nothing will happen.

Comment (Roxy-3D) [reply Marlin/Marlin_main.cpp:4092]: And just to make this a little bit more complex.   I think this issue only shows up when the printer homes to Z-Max, right?    We can probably make it so what ever 'fix' we put in place doesn't disrupt anybody else.

Comment (RowanMeara) [reply Marlin/Marlin_main.cpp:4092]: No, I believe that this is only an issue when the printer homes to Z-min.  On a home to Z-max the printer immediately starts moving in the homing direction.

**Comment by RowanMeara** [comment]
This issue has been hard to explain and created a lot of confusion in two different threads so far.  I think the whole reason that this is an issue in the first place is that the configuration file is incredibly unclear on how Z_HOMING_HEIGHT and Z_MAX_POS interract.  If people just set their Z_MAX_POS to a safe value, this would not be an issue.   I made some changes to try to make this very explicit in the configuration file.

**Comment by thinkyhead** [comment]
Let's see if simply setting `Z_HOMING_HEIGHT` to 0 is good enough for the setup in question. At bootup time there's no way to know where the bed and nozzle are, so every user has to decide on the assumptions that best fit their situation.

**Comment by RowanMeara** [comment]
Setting the Z_HOMING_HEIGHT to 0 does not fix this issue though.  If you are in negative Z and home, the Z axis will try to return to 0 which is completely unnecessary.  After startup, if you know that you have moved -50mm in the Z direction, I think it is safe to assume that you do not need to move all the way back to +4 (or whatever Z_HOMING_HEIGHT is set to) for a safe home.

I also think that the current documentation makes it difficult for the users to know what assumptions they are making.  There was a comment in the config file:
```
// Be sure you have this distance over your Z_MAX_POS in case.
```
I think that that comment was hard to understand and thus easy to ignore, and did not tell users exactly how Z_HOMING_HEIGHT and Z_MAX_POS are related.  I know that this comment would have helped me when I was configuring Marlin for the first time.
```
// To avoid crashes while homing the z-axis when starting from Z_MAX_POS,
// set Z_MAX_POS = (Physical z-axis length) - Z_HOMING_HEIGHT
```

**Comment by RowanMeara** [comment]
I also think that as Marlin expands CNC support, stuff like this is going to become critical.  People rarely move their 3d printers manually, but they manually control their CNCs all the time.  If Marlin is ever run on larger machines, crashing them will be costly rather than the minor annoyance it is today. 


 To me, the main issue here is that Z homing moves in the positive direction that are unexpectedly much larger than Z_HOMING_HEIGHT can occur.

**Comment by thinkyhead** [comment]
>  If you are in negative Z and home, the Z axis will try to return to 0 which is completely unnecessary.

Where is this negative Z position coming from?

**Comment by RowanMeara** [comment]
Before homing if you perform a G1 Z-50.  Obviously not a common use case currently, but I don't think we should ignore it.   The change to work around it is only four lines without the conditional compilation.  The conditional compilation is only to save a few bytes and I see now that it was a mistake to include it.

**Comment by thinkyhead** [comment]
> Before homing if you perform a G1 Z-50.

Before homing users shouldn't be moving things around blindly. We might have to enable this by default to stress the point…
```cpp
//#define NO_MOTION_BEFORE_HOMING  // Inhibit movement until all axes have been homed
```

**Comment by RowanMeara** [comment]
I completely agree that is something they should not be doing and that the config file should be updated to include this. 

However, I still think its worth fixing because the workaround does not add much overhead and is simple to understand and their are situations where this could be useful.

On a side note, I would support taking a close look at the main configuration file, and see if we can move more of the seldom used or unusual settings over to the advanced file.  A major pain point for many Marlin users is dealing with the enormous config file.  I think that we should at least move all development only options and see if we can condense it a bit

**Comment by thinkyhead** [comment]
I simply don't agree with the note in this change. It disregards those who home to Z max. And, I disagree that `Z_MAX_POS` should in any way relate to the `Z_HOMING_HEIGHT`, which is provided as a simple extra to prevent hitting the bed clips. Users will naturally know that when they start homing with a `Z_HOMING_HEIGHT` set what the behavior will necessarily be. The verbiage in this PR is narrow and misleading, and addresses an edge case that users are smart enough to figure out on their own.

> I would support taking a close look at the main configuration file, and see if we can move more of the seldom used or unusual settings over to the advanced file.

It occurs occasionally, and gradually. Filament Width Sensor was just moved, for example. In most cases it makes sense to keep sub-options close to their enable flag. If there are specific items that seem too complicated or obscure for the main config, feel free to submit PRs moving them, and each can be discussed on its own merits.

**Pull request closed without merging**